File: cpp/src/hypergraph.cpp
#include <iostream>
#include <vector>
#include <random>
#include <deque>
#include <set>
#include <unordered_map>
#include <algorithm>
#include "hypergraph.h"

HyperGraph::HyperGraph(){

	V = std::vector<int>();
	E = std::vector<int>();
	elist = std::unordered_map<int, std::vector<int>>();
	vlist = std::unordered_map<int, std::vector<int>>();
}

HyperGraph::~HyperGraph(){

	std::vector<int>().swap(V);
	std::vector<int>().swap(E);
	std::unordered_map<int, std::vector<int>>().swap(elist);
	std::unordered_map<int, std::vector<int>>().swap(vlist);
}

int HyperGraph::read_hypergraph(const char *hypergraph){

	const char *dir = "../../data/";

	FILE *f1;
	std::string f1path = std::string(dir) + hypergraph + "_nverts.txt";
	f1 = fopen(f1path.c_str(), "r");
	if(f1 == NULL) {
		printf("Error: Could not open file named %s_nverts.txt.\n", hypergraph);
		exit(0);
	}

	FILE *f2;
	std::string f2path = std::string(dir) + hypergraph + "_hyperedges.txt";
	f2 = fopen(f2path.c_str(), "r");
	if(f2 == NULL) {
		printf("Error: Could not open file named %s_hyperedges.txt.\n", hypergraph);
		exit(0);
	}

	int nv, v;
	int m = 0;
	V = std::vector<int>();
	E = std::vector<int>();
	elist = std::unordered_map<int, std::vector<int>>();
	vlist = std::unordered_map<int, std::vector<int>>();
	while(fscanf(f1, "%d", &nv) != EOF) {
		std::vector<int> e;
		for(int i=0; i<nv; ++i){
			auto fs = fscanf(f2, "%d", &v);
			if(fs != EOF){
				e.push_back(v);
			}
			else{
				std::string s = std::string(hypergraph) + "_hyperedges.txt";
				printf("Error: Failed to read a list of hyperedges named %s.\n", s.c_str());
				exit(0);
			}
		}

		vlist[m] = e;
		E.push_back(m);
		for(int v:e){
			if(elist.find(v) == elist.end()){
				elist[v] = std::vector<int>();
			}
			elist[v].push_back(m);

			if(std::find(V.begin(), V.end(), v) == V.end()){
				V.push_back(v);
			}
		}

		m += 1;
	}
	
	fclose(f1);
	fclose(f2);

	int N = int(V.size());
	int M = int(E.size());

	for(int i=0; i<N; ++i){
		std::sort(elist[i].begin(),elist[i].end());
	}
	for(int i=0; i<M; ++i){
		std::sort(vlist[i].begin(),vlist[i].end());
	}

	printf("The given hypergraph named %s was successfully read.\n", hypergraph);
	printf("Number of nodes: %d\n", N);
	printf("Number of hyperedges: %d\n\n", M);

	return 0;
}

int add_node_to_hyperedge(HyperGraph &G, const int &v, const int &m){
	if(G.elist.find(v) == G.elist.end()){
		printf("Error: Given node is not found.\n");
		exit(0);
	}
	if(G.vlist.find(m) == G.vlist.end()){
		printf("Error: Given hyperedge is not found.\n");
		exit(0);
	}

	G.vlist[m].push_back(v);
	std::sort(G.vlist[m].begin(),G.vlist[m].end());
	G.elist[v].push_back(m);
	std::sort(G.elist[v].begin(),G.elist[v].end());
	
	return 0;
}

int remove_node_from_hyperedge(HyperGraph &G, const int &v, const int &m){
	if(G.elist.find(v) == G.elist.end()){
		printf("Error: Given node is not found.\n");
		exit(0);
	}
	if(G.vlist.find(m) == G.vlist.end()){
		printf("Error: Given hyperedge is not found.\n");
		exit(0);
	}

	auto itr1 = std::find(G.elist[v].begin(), G.elist[v].end(), m);
	if(itr1 == G.elist[v].end()){
		printf("Error: Node %d is not included in hyperedge %d\n", v, m);
	}
	G.elist[v].erase(itr1);

	auto itr2 = std::find(G.vlist[m].begin(), G.vlist[m].end(), v);
	if(itr2 == G.vlist[m].end()){
		printf("Error: Node %d does not belong to hyperedge %d\n", v, m);
	}
	G.vlist[m].erase(itr2);

	return 0;
}

int calc_node_degree(HyperGraph &G, std::unordered_map<int, int> &node_degree){

	node_degree.clear();
	for(int v:G.V){
		node_degree[v] = int(G.elist[v].size());
	}

	return 0;
}

int calc_maximum_node_degree(HyperGraph &G, int &max_node_deg){
	int k;
	max_node_deg = 0;
	for(int v:G.V){
		k = int(G.elist[v].size());
		if(k > max_node_deg){
			max_node_deg = k;
		}
	}

	return 0;
}

int calc_num_jnt_node_deg(HyperGraph &G, Matrix<int> &num_jnt_node_deg){

	num_jnt_node_deg.clear();
	int s, i, j, u, v, k, l;
	for(int m:G.E){
		s = int(G.vlist[m].size());
		for(i=0; i<s-1; ++i){
			u = G.vlist[m][i];
			k = int(G.elist[u].size());
			for(j=i+1; j<s; ++j){
				v = G.vlist[m][j];
				l = int(G.elist[v].size());
				num_jnt_node_deg.add(k, l, 1);
				num_jnt_node_deg.add(l, k, 1);
			}
		}
	}

	return 0;
}

int calc_node_redundancy_coefficient(HyperGraph &G, std::unordered_map<int, double> &node_redun_coeff){
	node_redun_coeff.clear();
	int d, i, j, m1, m2;

	for(int v:G.V){
		d = int(G.elist[v].size());
		if(d < 2){continue;}

		double rc = 0.0;
		for(i=0; i<d-1; ++i){
			m1 = G.elist[v][i];
			for(j=i+1; j<d; ++j){
				m2 = G.elist[v][j];
				if(m1 == m2){continue;}
				std::set<int> s1, s3;
				std::set<int> s2 = {v};
				std::set_intersection(G.vlist[m1].begin(), G.vlist[m1].end(), G.vlist[m2].begin(), G.vlist[m2].end(), std::inserter(s1, s1.end()));
				std::set_difference(s1.begin(), s1.end(), s2.begin(), s2.end(), std::inserter(s3, s3.end()));

				if(int(s3.size()) > 0){
					rc += 2.0;
				}
			}
		}

		node_redun_coeff[v] = double(rc)/(d*(d-1));
	}

	return 0;
}

int calc_degree_dependent_node_redundancy_coefficient(HyperGraph &G, std::vector<double> &degree_node_redun_coeff){
	
	std::unordered_map<int, double> node_redun_coeff;
	calc_node_redundancy_coefficient(G, node_redun_coeff);

	int max_node_deg;
	calc_maximum_node_degree(G, max_node_deg);

	degree_node_redun_coeff = std::vector<double>(max_node_deg+1, 0.0);
	std::vector<int> N_k(max_node_deg+1, 0);

	for(int v:G.V){
		int k = int(G.elist[v].size());
		N_k[k] += 1;
		degree_node_redun_coeff[k] += node_redun_coeff[v];
	}

	for(int k=1; k<int(N_k.size()); ++k){
		if(N_k[k] > 0){
			degree_node_redun_coeff[k] = double(degree_node_redun_coeff[k])/N_k[k];
		}
	}

	return 0;
}

int calc_hyperedge_size(HyperGraph &G, std::unordered_map<int, int> &hyperedge_size){

	hyperedge_size.clear();
	for(int m:G.E){
		hyperedge_size[m] = int(G.vlist[m].size());
	}

	return 0;
}

File: py/hypergraph.py
import numpy as np
from collections import Counter
from collections import deque
from collections import defaultdict
import warnings
import hypernetx as hnx

class HyperGraph():
    N = 0
    M = 0
    E = []
    node_index = {}
    node_attribute = {}
    elist = {}

    def __init__(self):
        self.N = 0
        self.M = 0
        self.V = set()
        self.E = []
        self.node_index = {}
        self.node_attribute = {}
        self.elist = {}

    def nodes(self):

        return set(self.V)

    def hyperedges(self):

        return list(self.E)

    def add_node(self, v, attr: dict=None):
        if v not in self.node_index:
            self.node_index[v] = len(self.node_index)
            self.N = len(self.node_index)
            if attr is not None:
                self.node_attribute[v] = attr
            self.elist[v] = []

        self.V = set(self.node_index.keys())

        return

    def set_node_attribute(self, v, attr: dict):
        if v not in self.node_index:
            self.add_node(v, attr)
        else:
            self.node_attribute[v] = attr

        return

    def add_nodes_from(self, nodes, attrs: dict=None):
        for v in nodes:
            if attrs is None or attrs is not dict:
                self.add_node(v)
            elif v not in self.node_attribute:
                warnings.warn('Node attributes were inputted but node ' + str(v) + ' has no attribute ' + str(attrs) + '.')
                self.add_node(v)
            else:
                self.add_node(v, attrs[v])

        return

    def add_hyperedge(self, e: list):
        self.add_nodes_from(e)
        self.E.append(list(e))
        m = len(self.E) - 1
        for v in e:
            self.elist[v].append(m)
        self.M = len(self.E)

        return

    def add_hyperedges_from(self, hyperedges: list):
        nodes = set()
        for e in hyperedges:
            nodes = nodes | set(e)
        self.add_nodes_from(nodes)

        for e in hyperedges:
            self.add_hyperedge(e)

        return

    def add_node_to_hyperedge(self, v, m):
        # Add node v to hyperedge E[e_i]

        if v not in self.elist or m < 0 or m >= self.M:
            print("Error: Given node is not found.")
            exit()

        self.E[m].append(v)
        self.elist[v].append(m)

        return

    def remove_node_from_hyperedge(self, v, m):
        # Remove node v from hyperedge E[e_i]

        if v not in self.elist:
            print("Error: Given node is not found.")
            exit()
        if m < 0 or m >= self.M:
            print("Error: Given hyperedge is not found.")
            exit()

        if m not in self.elist[v]:
            warnings.warn('Given node is not included in the given hyperedge.')
        else:
            self.elist[v].remove(m)

        if v not in self.E[m]:
            warnings.warn('Given node does not belong to the given hyperedge.')
        else:
            self.E[m].remove(v)

        return

    def nodes_incident_with_hyperedge(self, m):
        if m < 0 or m >= self.M:
            print("ERROR: hyperedge index is out of range.")
            return []

        return self.E[m]

    def hyperedges_incident_with_node(self, v):
        if v not in self.node_index:
            print("ERROR: node index is out of range.")
            return []

        return self.elist[v]

    def is_connected(self):

        searched = {i: 0 for i in self.node_index}
        nodes = set()
        v = np.random.choice(list(self.node_index.keys()))

        Q = deque()
        searched[v] = 1
        Q.append(v)
        while len(Q) > 0:
            v = Q.popleft()
            nodes.add(v)
            for m in self.elist[v]:
                for w in self.E[m]:
                    if searched[w] == 0:
                        searched[w] = 1
                        Q.append(w)

        return self.N == sum(list(searched.values()))

    def node_degree(self, v=None, ndarray=False):
        if v is None:
            if ndarray is False:
                node_degree = {}
                for m in range(0, self.M):
                    for v in self.E[m]:
                        node_degree[v] = node_degree.get(v, 0) + 1
                return node_degree
            else:
                node_degree = np.zeros(self.N, dtype=int)
                for m in range(0, self.M):
                    for v in self.E[m]:
                        node_degree[v] += 1
                return node_degree
        elif v not in self.elist:
            print("ERROR: node index is out of range.")
            return -1
        else:
            return len(self.elist[v])

    def hyperedge_size(self, m=None, ndarray=False):
        if m is None:
            if ndarray is False:
                return {m: len(self.E[m]) for m in range(0, self.M)}
            else:
                l = np.zeros(self.M, dtype=int)
                for m in range(0, self.M):
                    l[m] = len(self.E[m])
                return l
        elif m < 0 or m >= self.M:
            print("ERROR: hyperedge index is out of range.")
            return -1
        else:
            return len(self.E[m])

    def node_degree_distribution(self, func):
        if func == 'freq':
            return dict(Counter(list(self.node_degree().values())))
        elif func == 'prob':
            d = dict(Counter(list(self.node_degree().values())))
            return {k: float(d[k])/self.N for k in d}
        elif func == 'survival':
            d_ = dict(Counter(list(self.node_degree().values())))
            lst = sorted(list(d_.items()), key=lambda x:x[0], reverse=False)
            d = {}
            for i in range(0, len(lst)):
                k = lst[i][0]
                n = sum([lst[j][1] for j in range(i, len(lst))])
                d[k] = float(n)/self.N
            return d
        else:
            print("ERROR: given function is not supported.")
            return {}

    def hyperedge_size_distribution(self, func):
        if func == 'freq':
            return dict(Counter(list(self.hyperedge_size().values())))
        elif func == 'prob':
            d = dict(Counter(list(self.hyperedge_size().values())))
            return {k: float(d[k])/self.N for k in d}
        elif func == 'survival':
            d_ = dict(Counter(list(self.hyperedge_size().values())))
            lst = sorted(list(d_.items()), key=lambda x:x[0], reverse=False)
            d = {}
            for i in range(0, len(lst)):
                k = lst[i][0]
                n = sum([lst[j][1] for j in range(i, len(lst))])
                d[k] = float(n)/self.M
            return d
        else:
            print("ERROR: given function is not supported.")
            return {}

    def joint_node_degree_matrix(self, ndarray=False):
        # Calculate the number of hyperedges that nodes with degree k and nodes with degree k' share.

        node_degree = self.node_degree()
        node_degrees = set(list(node_degree.values()))

        jnd = {(k1, k2): 0 for k1 in node_degrees for k2 in node_degrees}

        for e in self.E:
            s = int(len(e))
            for i in range(0, s - 1):
                u = e[i]
                k1 = int(len(self.elist[u]))
                for j in range(i + 1, s):
                    v = e[j]
                    k2 = int(len(self.elist[v]))
                    jnd[(k1,k2)] += 1
                    jnd[(k2,k1)] += 1

        if ndarray:
            max_deg = max(node_degrees)
            jnd_ = np.zeros((max_deg+1, max_deg+1), dtype=float)
            for (k, l) in jnd:
                jnd_[k][l] += jnd[(k, l)]
            return jnd_
        else:
            return jnd

    def joint_node_degree_distribution(self, ndarray=False):

        node_degree = self.node_degree()
        node_degrees = set(list(node_degree.values()))

        jnd = {(k1, k2): 0 for k1 in node_degrees for k2 in node_degrees}
        norm = 0

        for e in self.E:
            s = int(len(e))
            norm += s * (s - 1)
            for i in range(0, s - 1):
                u = e[i]
                k1 = int(len(self.elist[u]))
                for j in range(i + 1, s):
                    v = e[j]
                    k2 = int(len(self.elist[v]))
                    jnd[(k1,k2)] += 1
                    jnd[(k2,k1)] += 1

        if ndarray:
            max_deg = max(node_degrees)
            jnd_ = np.zeros((max_deg+1, max_deg+1), dtype=float)
            for (k, l) in jnd:
                jnd_[k][l] = float(jnd[(k, l)]) / norm
            return jnd_
        else:
            for (k, l) in jnd:
                jnd[(k,l)] /= norm
            return jnd

    def average_degree_of_nearest_neighbors_of_nodes(self):
        sum_neighbor_node_degree = {}
        N_k = {}
        node_degree = self.node_degree()

        for v in self.nodes():
            k = node_degree[v]
            N_k[k] = N_k.get(k, 0) + 1

            x = 0
            y = 0
            for m in self.elist[v]:
                e = self.E[m]
                for w in e:
                    l = node_degree[w]
                    x += 1
                    y += l
                x -= 1
                y -= k
            if x > 0:
                sum_neighbor_node_degree[k] = sum_neighbor_node_degree.get(k, 0) + float(y) / x

        annd = defaultdict(float)
        for k in N_k:
            if N_k[k] > 0:
                annd[k] = float(sum_neighbor_node_degree[k]) / N_k[k]

        return annd

    def node_redundancy_coefficient(self, ndarray=False):
        # Calculate the redundancy coefficient of each node.
        # See the following paper for the detail of the redundancy coefficient.
        # Basic notions for the analysis of large two-mode networks, M. Latapy, C. Magnien, N. Del Vecchio, Social networks, 2008.

        nodes = self.nodes()
        rc = {v: 0.0 for v in nodes}

        for v in nodes:
            d = len(self.elist[v])

            if d < 2:
                rc[v] = 0
                continue

            for i in range(0, d - 1):
                e1 = self.elist[v][i]
                for j in range(i + 1, d):
                    e2 = self.elist[v][j]
                    if e1 == e2:
                        continue
                    s = set(set(set(self.E[e1]) & set(self.E[e2])) - set(list([v])))
                    if len(s) > 0:
                        rc[v] += 2.0

            rc[v] = float(rc[v]) /(d * (d - 1))

        if ndarray:
            rc_ = np.zeros(self.N, dtype=float)
            for v in nodes:
                rc_[v] = rc[v]
            return rc_
        else:
            return rc

    def degree_dependent_node_redundancy_coefficient(self, ndarray=False):
        # Calculate the degree-dependent redundancy coefficient of the node (i.e., the average of the redundancy coefficient over the nodes with degree k).

        rc = self.node_redundancy_coefficient()
        nd = self.node_degree()

        node_degrees = set()
        nodes = self.nodes()
        for v in nodes:
            k = int(len(self.elist[v]))
            node_degrees.add(k)

        ddrc = {k: 0.0 for k in node_degrees}
        n_k = {k: 0.0 for k in node_degrees}

        for v in nodes:
            k = nd[v]
            ddrc[k] += rc[v]
            n_k[k] += 1

        for k in ddrc:
            if n_k[k] > 0:
                ddrc[k] = float(ddrc[k]) / n_k[k]

        if ndarray:
            max_d = max(node_degrees)
            ddrc_ = np.zeros(max_d+1, dtype=float)
            for k in ddrc:
                ddrc_[k] = ddrc.get(k, 0)
            return ddrc_
        else:
            return ddrc

    def node_shortest_path_length_distribution(self):
        dist = {}

        for s in self.node_index:
            d = {w: -1 for w in self.node_index}
            Q = deque()
            d[s] = 0
            Q.append(s)

            while len(Q) > 0:
                v = Q.popleft()
                for m in self.elist[v]:
                    for w in self.E[m]:
                        if d[w] < 0:
                            d[w] = d[v] + 1
                            Q.append(w)

            for w in d:
                if w == s or d[w] <= 0:
                    continue
                k = d[w]
                dist[k] = dist.get(k, 0) + 1

        return dist

    def single_source_node_shortest_path_length(self, v):
        if v not in self.node_index:
            print("ERROR: given node is not found.")
            return {}

        d = {w: -1 for w in self.node_index}

        s = v
        Q = deque()
        d[s] = 0
        Q.append(s)

        while len(Q) > 0:
            v = Q.popleft()
            for m in self.elist[v]:
                for w in self.E[m]:
                    if d[w] < 0:
                        d[w] = d[v] + 1
                        Q.append(w)

        return d

    def rich_club_coefficient(self):
        node_degrees = set()
        for v in self.node_index:
            k = int(len(self.elist[v]))
            node_degrees.add(k)

        min_k, max_k = 1, max(node_degrees)
        rcc = {k: 0 for k in range(min_k, max_k + 1)}
        for e in self.E:
            k = min([self.node_degree(v) for v in e])
            for k_ in range(1, k):
                rcc[k_] += 1

        return rcc

    def print_info(self):
        node_degree = {i: 0 for i in self.node_index}
        hyperedge_size = {m: 0 for m in range(0, self.M)}
        for m in range(0, self.M):
            for i in self.E[m]:
                node_degree[i] += 1
            hyperedge_size[m] = len(self.E[m])

        node_degree_lst = list(node_degree.values())
        hyperedge_size_lst = list(hyperedge_size.values())

        print("Number of nodes:", self.N)
        print("Number of hyperedges:", self.M)
        print("Average degree of node:", float(sum(node_degree_lst)) / self.N)
        print("Maximum degree of node:", max(node_degree_lst))
        print("Frequency distribution of node degree:", dict(Counter(node_degree_lst)))
        print("Average size of hyperedge:", float(sum(hyperedge_size_lst)) / self.M)
        print("Maximum size of hyperedge:", max(hyperedge_size_lst))
        print("Frequency distribution of hyperedge size:", dict(Counter(hyperedge_size_lst)))
        print("Hypergraph is connected:", self.is_connected())
        print()

        return

    def convert_to_hnx_hypergraph(self):
        B = np.zeros((self.N, self.M), dtype=int)
        for m in range(0, self.M):
            for v in self.E[m]:
                B[v][m] += 1

        H_hnx = hnx.Hypergraph.from_incidence_matrix(B)

        return H_hnx

def read_hypergraph(hypergraph_name):
    # Read hypergraph named hypergraph_name.
    # The corresponding files to read must be in the folder ./hyper-dk-series/data/.

    f1_path = "../data/" + str(hypergraph_name) + "_nverts.txt"
    f1 = open(f1_path, 'r')
    f2_path = "../data/" + str(hypergraph_name) + "_hyperedges.txt"
    f2 = open(f2_path, 'r')

    lines1 = f1.readlines()
    lines2 = f2.readlines()

    E = []
    c = 0
    for line1 in lines1:
        nv = int(line1[:-1].split(" ")[0])

        e = []
        for i in range(0, nv):
            v = int(lines2[c+i][:-1])
            e.append(v)

        E.append(e)
        c += nv

    f1.close()
    f2.close()

    H = HyperGraph()
    H.add_hyperedges_from(E)

    return H

def write_hypergraph(H: HyperGraph, hypergraph_name):
    f1_path = "../data/" + str(hypergraph_name) + "_nverts.txt"
    f1 = open(f1_path, 'w')

    for e in H.E:
        f1.write(str(int(len(e))) + "\n")

    f1.close()

    f2_path = "../data/" + str(hypergraph_name) + "_hyperedges.txt"
    f2 = open(f2_path, 'w')

    for e in H.E:
        for v in e:
            f2.write(str(int(v)) + "\n")

    f2.close()

    # print("Wrote the " + str(k) + "-th hypergraph randomized with (" + d_v + ", " + d_e + ").\n")

    return
